---
title: "Group 6 Project"
author: "Joowon Lee"
date: "11/15/2019"
output:
  word_document: default
  pdf_document: default
  html_document:
    df_print: paged
editor_options:
  chunk_output_type: console
---

Notes:
  - You do not have to put all of your team members' code into a single file. I have included all 5 analyses just for your information. You only need the code for your analysis.
  - The tasks include both coding and written interpretation.
  - Please knit to word document -- it will make it easier to combine your results with your team members in to the single manuscript (submitted in GP4).

## Setup

### Load packages

```{r setup, include=FALSE}
### Global knitr options
knitr::opts_chunk$set(echo = TRUE)

### Load packages/libraries that we will need
library(tidyverse)
library(viridis)    # Better plot colors
library(cluster)    # clustering algorithms
library(factoextra) # clustering algorithms & visualization
library(ClusterR)
library(useful)
```

### Custom ggplot theme

```{r theme}

### DON'T EDIT CODE IN THIS CHUNK

theme_custom <- theme_bw() +

  # if we have a plot title or subtitle, let's center it
  theme (
    plot.title = element_text(hjust = 0.5),
    plot.subtitle = element_text(hjust = 0.5)
  )
theme_set(theme_custom)

### We'll make the viridis color scale our default plotting color palette
scale_colour_continuous <- function(...) {
  scale_colour_viridis_c(...)
}
scale_fill_continuous <- function(...) {
  scale_fill_viridis_c(...)
}
scale_colour_discrete <- function(...) {
  scale_colour_viridis_d(..., begin = 0, end = 0.9)
}
scale_fill_discrete <- function(...) {
  scale_fill_viridis_d(..., begin = 0, end = 0.9)
}
```

### Setwd fix (if needed)

If you are having trouble loading the exprs_tidy file below, manually override the working directory. To do this
  1. In the menu bar, click: Session > Set Working Directory > To Source File Location
  2. Copy the line of code in the console, and paste it into the code chunk below

```{r fix_setwd}
### EDIT if necessary
setwd("~/Desktop")
```

### FYI: how I got the data

```{r get_data, eval=FALSE}
### Get list of available datasets
### https://www.bioconductor.org/packages/3.3/bioc/vignettes/TCGAbiolinks/inst/doc/tcgaBiolinks.html#harmonized-data-1
View(getGDCprojects())

### Datasets to use for group project (I picked the ones with smallest sample size and no sex bias)
projects <- c(
  "TCGA-ACC",
  "TCGA-CHOL",
  "TCGA-DLBC",
  "TCGA-KICH",
  "TCGA-MESO",
  "TCGA-UVM"
)

phenoList <-  vector(mode = "list", length = length(projects))
names(phenoList) <- projects
exprsList <-  vector(mode = "list", length = length(projects))
names(exprsList) <- projects
for (i in projects) {
  ### Get data (in summarized experiment ["se"]  format)
  query <- GDCquery(
    project = i,
    data.category = "Transcriptome Profiling",
    data.type = "Gene Expression Quantification",
    workflow.type = "HTSeq - FPKM"
  )
  GDCdownload(query)
  se <- GDCprepare(query)

  ### Extract phenoData and remove columns that either are all different or all consistent
  pheno_full <- as.data.frame(colData(se))
  pheno <- janitor::remove_constant(pheno_full)

  ### Extract exprs matrix and remove lowly expressed
  exprs_full <- assay(se)
  keep <- rowSums(exprs_full > 1) >= 10
  exprs <- exprs_full[keep, ]

  ### Shorten the sample id
  rownames(pheno) <- abbreviate(gsub("TCGA-OR-", "", rownames(pheno)), method = "both")
  pheno$id <- rownames(pheno)
  colnames(exprs) <- abbreviate(gsub("TCGA-OR-", "", colnames(exprs)), method = "both")

  ### Remove extra columns (not groups)
  pheno$sample <- pheno$id
  pheno$id <- NULL
  remove_cols <- c(
    "patient", "updated_datetime", "updated_datetime.x", "updated_datetime.y",
    "barcode", "diagnosis_id", "demographic_id", "exposure_id", "bcr_patient_barcode",
    "morphology", "treatments",
    "days_to_birth", "days_to_last_follow_up", "days_to_death",
    "year_of_birth", "year_of_diagnosis", "year_of_death"
  )
  pheno <- pheno[ , !(colnames(pheno) %in% remove_cols)]
  pheno <- pheno[ , !(colnames(pheno) %in% colnames(pheno)[grep("_CHOL_del|_CHOL_amp|subtype_", colnames(pheno))])]

  ### Save
  saveRDS(exprs, paste0(i, "_exprs.rds"))
  saveRDS(pheno, paste0(i, "_pheno.rds"))

  ### Add to list
  exprsList[[i]]  <- exprs
  phenoList[[i]] <- pheno

  ### Clean up
  rm(exprs)
  rm(exprs_full)
  rm(pheno)
  rm(pheno_full)
  rm(keep)
}

### Save
saveRDS(exprsList, "all_exprs.rds")
saveRDS(phenoList, "all_pheno.rds")

### Look at
sapply(exprsList, dim)
sapply(phenoList, dim)
sapply(phenoList, names)

### Write out names
rbind(
  paste("ACC:", toString(sort(names(phenoList$`TCGA-ACC`)))),
  paste("CHOL:", toString(sort(names(phenoList$`TCGA-CHOL`)))),
  paste("DLBC:", toString(sort(names(phenoList$`TCGA-DLBC`)))),
  paste("KICH:", toString(sort(names(phenoList$`TCGA-KICH`)))),
  paste("MESO:", toString(sort(names(phenoList$`TCGA-MESO`)))),
  paste("UVM:", toString(sort(names(phenoList$`TCGA-UVM`))))
) %>%
  writeLines("sample_variables.txt")
```

## Pre-process data

Your entire team should use the same code for this section!

### Load your dataset

```{r load_data}
### EDIT: You need to insert your dataset file names in the quotes below
exprs <- readRDS("TCGA-UVM_exprs.rds")
pheno <- readRDS("TCGA-UVM_pheno.rds")
```

### Fix sample names

```{r}
colnames(exprs) <- gsub("-", "_", colnames(exprs))
rownames(pheno) <- gsub("-", "_", rownames(pheno))
pheno$sample <- rownames(pheno)
```

### Pick your group (variable of interest) [edit!]

This should be a variable that is categorical with at least 2 categories and at least 3 samples in each category Use colnames(pheno) to find out what variable options you have. You can use one of the descriptive summary functions (from AE3) to determine how many categories there are for each group, and how many samples there are for each category.

```{r select_group}
### EDIT!! Copy your variable of interest into a new column called "group". This will help generalize/simplify your project's code

pheno$group <- pheno$age_at_index

### I grouped ages the same for all Team 6 members
pheno$group <- ifelse(
  pheno$age_at_index < 55, "<55", ifelse(
    pheno$age_at_index > 70, ">70", "55-70"
  )
)
```

### Filter samples

Check for samples with missing data for your "group"

```{r filter_samples_check}
### Don't edit

### You can check this using the following (many other ways to check too)
### Make sure no "blanks" either --
### sometimes missing data isn't converted to an NA but instead is just blank
summary(as.factor(pheno$group)) # look for blanks, no data, etc. categories
table(is.na(pheno$group))
```

Remove samples with missing data (or no data; i.e. "not reported") for your "group". First from phenoData

```{r filter_samples_remove}
### Don't edit

### Remove NAs
pheno <- pheno[!is.na(pheno$group), ]

### Remove blanks
pheno <- pheno[!(pheno$group == ""), ]

### Remove "not reported"
pheno <- pheno[!(pheno$group == "not reported"), ]
```

And also from exprsData

```{r}
### Don't edit

exprs <- exprs[ , which(colnames(exprs) %in% rownames(pheno))]
```

### Filter genes

Low expression: Here we'll require FPKM > 5 in at least 25 samples (it will help reduce computational time)

```{r filter_genes}
### Don't edit

### Filter
exprs <- exprs[rowSums(exprs > 5) > 25, ]
```

### Finalize data

Match order of samples in both pheno (rownames) and exprs (colnames) datasets

```{r matrices}
### Don't edit

### Match order of samples in both matrices
table(colnames(exprs) == rownames(pheno)) # check
pheno <- pheno[match(colnames(exprs), rownames(pheno)), ] # fixes if needed
table(colnames(exprs) == rownames(pheno)) # check
```

Convert expression matrix to tidy, so that we also have a tidy version of the data in case we need it.

```{r tidy}
### Don't edit

### Create tibble with expression and pheno data
tidy <- exprs %>%

  # Convert expression matrix to tidy tibble
  as_tibble(rownames = "gene") %>%
  gather("sample", "fpkm", -gene)  %>%

  # Add phenoData
  inner_join(pheno, by = "sample")
```

## Team Member #5: K-means clustering

Tasks:
  a. Determine ideal clustering methods (including k value)
  b. Generate a final figure of the clustering
  c. Provide table summarizing the number of samples in each cluster and the breakdown of samples by your group of interest
  d. Interpret is the clustering pattern in relation to your group of interest

### Scaling

```{r detecting na and scaling}
### Transpose so clustering samples
exprs_mat <- t(exprs)

### Scaling so that the data can be clustered
set.seed(1234)
exprs_mat <- scale(exprs_mat)
```

### Determine the optimal k value

Method 1: WSS ("Elbow") method

```{r kmeans_wss}
set.seed(1234) 
fviz_nbclust(exprs_mat, kmeans, method = "wss")
```
```{r}
Sample_age_at_index <- FitKMeans(exprs_mat, max.clusters = 10, nstart = 25, seed = 1234)
Sample_age_at_index
PlotHartigan(Sample_age_at_index)
```

Method 2: Gap Statistic Method

```{r kmeans_gap}
### It takes awhile at the bootstrapping step. To help reduce the time, I reduced the number of bootstraps from 100 to 25 by using "nboot = 25". 

set.seed(1234)
fviz_nbclust(exprs_mat, kmeans, nboot = 25, method = "gap_stat")
```
```{r}
Gaps <- clusGap(exprs_mat, FUNcluster = kmeans, K.max = 10)
gap_stat <- as.data.frame(Gaps$Tab)
gap_stat

# logW curves
ggplot(gap_stat, aes(x=1:nrow(gap_stat))) +
  geom_line(aes(y=logW), color ="blue") +
  geom_point(aes(y=logW), color ="blue") +
  geom_line(aes(y=E.logW), color ="green") +
  geom_point(aes(y=E.logW), color ="green") +
  labs(x="Number of Clusters", y="logW")

# gap curve
ggplot(gap_stat, aes(x=1:nrow(gap_stat))) +
  geom_line(aes(y=gap), color ="red") +
  geom_point(aes(y=gap), color ="red") +
  geom_errorbar(aes(ymin=gap-SE.sim, ymax=gap+SE.sim), color="red") +
  labs(x="Number of Clusters", y="Gap")
```

Optimal K-Value

```{r kmeans_final_k}
k_value <- 2
```

### Final k-means run using optimal k value

```{r kmeans_final}
### Run
set.seed(1234)
km_final <- kmeans(exprs_mat, k_value, nstart = 25)

### Plot
p <- fviz_cluster(
  km_final,
  data = exprs_mat,
  geom = "text",
  labelsize = 5, repel = TRUE
) +

  # Formatting
  theme_classic() +
  ggtitle(paste("k =", k_value)) +

  # Add space to axes, so that you can read the sample names
  scale_x_continuous(expand = c(0.2, 0.2)) +
  scale_y_continuous(expand = c(0.2, 0.2))

### Print
p

### Save
# From Alexis: added this
ggsave("Lee_Joowon_kmeans_plot.png")
```

### Add cluster assignments to phenoData

```{r}
### Add columns for HC and k-means clusters
res_clusters <- pheno %>%
  mutate(
    kmeans_cluster = paste0("cluster_", km_final$cluster)
  )
```

### Summarize cluster assignments by group

As table

```{r}
### Get absolute number of samples in each cluster by group
res_clusters %>%
  select(group, kmeans_cluster) %>% 
  table()
```

As figure

```{r}
### Visualize percent of samples in each cluster by group
res_clusters %>%
  
  # Get percent of samples in each group in each cluster
  dplyr::group_by(group, kmeans_cluster) %>% 
  dplyr::summarise(n = n()) %>%
  spread(kmeans_cluster, n) %>%
  mutate(n_samples = sum(cluster_1, cluster_2, na.rm = TRUE)) %>%
  mutate(
    cluster_1 = round(100*cluster_1/n_samples),
    cluster_2 = round(100*cluster_2/n_samples)
  ) %>%
  select(group, n_samples, dplyr::everything()) %>%
  gather(cluster, percent, -group, -n_samples) %>%
  
  # Plot
  ggplot(aes(x = cluster, y = group, fill = percent)) +
  geom_tile() +
  
  # Formatting
  theme_classic() +
  ggtitle(paste0("Clustering by Group ", "(k = ", k_value, ")")) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0))
```

## Session info

```{r sessioninfo}
sessionInfo()
```
